File: app/src/test/cpp/NativeAudioEffectsController.java
package top.nekoh2o.player.audio;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

// Host JNI contract test. Keep declarations aligned with the Kotlin controller.
public final class NativeAudioEffectsController {
    native long nativeCreate(int rate, int channels);
    native void nativeConfigure(long handle, float[] bands, int bass, int width, int wet,
        int room, int damping, int loudness, int masteringId, int masteringMix);
    native void nativeProcess(long handle, ByteBuffer buffer, int samples);
    native void nativeReset(long handle);
    native void nativeRelease(long handle);

    private static void expectFailure(Runnable operation) {
        try { operation.run(); }
        catch (IllegalStateException expected) { return; }
        throw new AssertionError("JNI should reject invalid input");
    }

    public static void main(String[] args) {
        System.load(args[0]);
        NativeAudioEffectsController controller = new NativeAudioEffectsController();
        expectFailure(() -> controller.nativeCreate(44100, 0));
        expectFailure(() -> controller.nativeCreate(0, 2));
        for (int channels : new int[] {1, 2}) {
            long handle = controller.nativeCreate(48000, channels);
            if (handle == 0) throw new AssertionError("No DSP instance");
            try {
                controller.nativeConfigure(handle, new float[10], 0, 0, 0, 50, 30, 0, 0, 100);
                ByteBuffer buffer = ByteBuffer.allocateDirect(8).order(ByteOrder.nativeOrder());
                buffer.putShort(0, (short)-32768);
                buffer.putShort(2, (short)32767);
                buffer.putShort(4, (short)123);
                buffer.putShort(6, (short)-123);
                controller.nativeProcess(handle, buffer, 4);
                if (buffer.getShort(0) != -32768 || buffer.getShort(2) != 32767 ||
                    buffer.getShort(4) != 123 || buffer.getShort(6) != -123)
                    throw new AssertionError("Bypass changed PCM");
                expectFailure(() -> controller.nativeProcess(handle, buffer, 5));
                expectFailure(() -> controller.nativeProcess(handle, ByteBuffer.allocate(8), 4));
                expectFailure(() -> controller.nativeProcess(handle, null, 4));
                expectFailure(() -> controller.nativeConfigure(handle, new float[9], 0, 0, 0, 0, 0, 0, 0, 100));
                for (int preset = 15; preset <= 34; preset++) {
                    controller.nativeConfigure(handle, new float[10], 0, 0, 0, 50, 30, 0, preset, 100);
                    controller.nativeProcess(handle, buffer, 4);
                    controller.nativeReset(handle);
                }
            } finally { controller.nativeRelease(handle); }
        }
        System.out.println("JNI checks passed: exports, mono/stereo, exact bypass, 20 presets, buffer bounds, lifecycle");
    }
}

File: app/src/main/java/top/nekoh2o/player/audio/NativeAudioEffectsController.kt
package top.nekoh2o.player.audio

import java.nio.ByteBuffer
import top.nekoh2o.player.data.model.AudioEffectSettings

/** 每个 AudioProcessor 独占一个 DSP 实例，仅在播放线程调用。 */
internal interface NativeEffectBackend {
    fun create(sampleRate: Int, channelCount: Int): Long
    fun configure(handle: Long, settings: AudioEffectSettings)
    fun process(handle: Long, buffer: ByteBuffer, sampleCount: Int)
    fun reset(handle: Long)
    fun release(handle: Long)
}

internal class JniEffectBackend : NativeEffectBackend {
    override fun create(sampleRate: Int, channelCount: Int): Long =
        NativeAudioEffectsController.nativeCreate(sampleRate, channelCount)

    override fun configure(handle: Long, settings: AudioEffectSettings) {
        NativeAudioEffectsController.nativeConfigure(
            handle, settings.eqBands.toFloatArray(), settings.bassBoost,
            settings.virtualizer, settings.reverbWet, settings.reverbRoomSize,
            settings.reverbDamping, settings.loudnessGain,
            settings.masteringPresetId, settings.masteringMix
        )
    }

    override fun process(handle: Long, buffer: ByteBuffer, sampleCount: Int) =
        NativeAudioEffectsController.nativeProcess(handle, buffer, sampleCount)
    override fun reset(handle: Long) = NativeAudioEffectsController.nativeReset(handle)
    override fun release(handle: Long) = NativeAudioEffectsController.nativeRelease(handle)
}

internal object NativeAudioEffectsController {
    init { System.loadLibrary("nekoplayer_audio_effects") }

    external fun nativeCreate(sampleRate: Int, channelCount: Int): Long
    external fun nativeConfigure(
        handle: Long, bands: FloatArray, bass: Int, width: Int, wet: Int,
        room: Int, damping: Int, loudness: Int, masteringId: Int, masteringMix: Int
    )
    external fun nativeProcess(handle: Long, buffer: ByteBuffer, sampleCount: Int)
    external fun nativeReset(handle: Long)
    external fun nativeRelease(handle: Long)
}
